src/runtime/pkg/kata-monitor/cri.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2020 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package katamonitor
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/url"
+	"regexp"
+
+	"github.com/kata-containers/kata-containers/src/runtime/pkg/types"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/xeipuuv/gojsonpointer"
+	"google.golang.org/grpc"
+
+	pb "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+const (
+	// unixProtocol is the network protocol of unix socket.
+	unixProtocol = "unix"
+)
+
+// getAddressAndDialer returns the address parsed from the given endpoint and a context dialer.
+func getAddressAndDialer(endpoint string) (string, func(ctx context.Context, addr string) (net.Conn, error), error) {
+	protocol, addr, err := parseEndpointWithFallbackProtocol(endpoint, unixProtocol)
+	if err != nil {
+		return "", nil, err
+	}
+	if protocol != unixProtocol {
+		return "", nil, fmt.Errorf("only support unix socket endpoint")
+	}
+
+	return addr, dial, nil
+}
+
+func getConnection(endPoint string) (*grpc.ClientConn, error) {
+	var conn *grpc.ClientConn
+	monitorLog.Debugf("connect using endpoint '%s' with '%s' timeout", endPoint, defaultTimeout)
+	addr, dialer, err := getAddressAndDialer(endPoint)
+	if err != nil {
+		return nil, err
+	}
+	conn, err = grpc.Dial(addr, grpc.WithInsecure(), grpc.WithBlock(), grpc.WithTimeout(defaultTimeout), grpc.WithContextDialer(dialer))
+	if err != nil {
+		errMsg := errors.Wrapf(err, "connect endpoint '%s', make sure you are running as root and the endpoint has been started", endPoint)
+		return nil, errMsg
+	}
+	monitorLog.Debugf("connected successfully using endpoint: %s", endPoint)
+	return conn, nil
+}
+
+func matchesRegex(pattern, target string) bool {
+	if pattern == "" {
+		return true
+	}
```

Comment:
This feels like it might cause problems. How about making this function return an `error`: that way, if it returns `nil`, the match was successful, otherwise we can return a detailed error message explaining why.

Related: it would be good to have a few unit tests for these functions.

Reply:
It looked better to me to rework that part a bit: removed that function and tried to make the code simpler and more clear: the idea is that we can just check if a "kata" substring is present in the lowRuntime var, no matter which container engine we use.
Moreover, if we don't get the runtime from a Pod that means that either we are using a newer container engine or a newer version that changed the metadata associated with the pod. This case deserves to be logged at least at the Info level.

Yeah, will see what unit tests to add.